components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java

```diff
@@ -808,7 +825,240 @@ public ConnectedApps getConnectedApps(String resourceId, Integer limit, Integer
         }
     }
 
-    // Private utility Methods.
+    /**
+     * Retrieve the list of IDP templates.
+     *
+     * @param limit      Items per page.
+     * @param offset     Offset.
+     * @param filter     Filter string. E.g. filter="name" sw "google" and "isEnabled" eq "true"
+     * @return  List of identity templates.
+     */
+    public IdentityProviderTemplateListResponse getIDPTemplates(Integer limit, Integer offset, String filter) {
+
+        try {
+            TemplateManager templateManager = IdentityProviderServiceHolder.getTemplateManager();
+            List<Template> templateList = templateManager.listTemplates(
+                    TemplateMgtConstants.TemplateType.IDP_TEMPLATE.toString(), limit, offset);
+            return createIDPTemplateListResponse(templateList, offset, limit, filter);
+        } catch (TemplateManagementException e) {
+            throw handleTemplateMgtException(e, Constants.ErrorMessage.ERROR_CODE_ERROR_LISTING_IDP_TEMPLATES, null);
+        }
+    }
+
+    /**
+     * Get an identity provider template identified by resource ID.
+     *
+     * @param templateId IDP template Id
+     * @return IdentityProviderTemplateResponse
+     */
+    public IdentityProviderTemplate getIDPTemplate(String templateId) {
+
+        try {
+            Template idpTemplate = IdentityProviderServiceHolder.getTemplateManager().getTemplateById(templateId);
```

Comment:
check whether idpTemplate is null. If null, send 404 not found.

Reply:
TemplateManager throws an exception when resource is not found. So this is handled at handleTemplateMgtException method. Is it necessary to handle the null in here also?

Reply:
Then it is fine. However you need to map the error code of that exception to http 404 in https://github.com/wso2/identity-rest-dispatcher/blob/master/components/org.wso2.carbon.identity.api.dispatcher/src/main/resources/ErrorMappings.properties